File: postgresql-async/src/main/scala/com/github/mauricio/async/db/postgresql/column/PostgreSQLColumnEncoderRegistry.scala
/*
 * Copyright 2013 Maurício Linhares
 *
 * Maurício Linhares licenses this file to you under the Apache License,
 * version 2.0 (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at:
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */

package com.github.mauricio.async.db.postgresql.column

import java.nio.ByteBuffer

import com.github.mauricio.async.db.column._
import io.netty.buffer.ByteBuf
import org.joda.time._

import scala.collection.JavaConversions._

object PostgreSQLColumnEncoderRegistry {
  val Instance = new PostgreSQLColumnEncoderRegistry()
}

class PostgreSQLColumnEncoderRegistry extends ColumnEncoderRegistry {

  private val classesSequence_ : List[(Class[_], (ColumnEncoder, Int))] = List(
    classOf[Int] -> (IntegerEncoderDecoder -> ColumnTypes.Integer),
    classOf[java.lang.Integer] -> (IntegerEncoderDecoder -> ColumnTypes.Integer),

    classOf[java.lang.Short] -> (ShortEncoderDecoder -> ColumnTypes.Smallint),
    classOf[Short] -> (ShortEncoderDecoder -> ColumnTypes.Smallint),

    classOf[Long] -> (LongEncoderDecoder -> ColumnTypes.Bigserial),
    classOf[java.lang.Long] -> (LongEncoderDecoder -> ColumnTypes.Bigserial),

    classOf[String] -> (StringEncoderDecoder -> ColumnTypes.Varchar),
    classOf[java.lang.String] -> (StringEncoderDecoder -> ColumnTypes.Varchar),

    classOf[Float] -> (FloatEncoderDecoder -> ColumnTypes.Real),
    classOf[java.lang.Float] -> (FloatEncoderDecoder -> ColumnTypes.Real),

    classOf[Double] -> (DoubleEncoderDecoder -> ColumnTypes.Double),
    classOf[java.lang.Double] -> (DoubleEncoderDecoder -> ColumnTypes.Double),

    classOf[BigDecimal] -> (BigDecimalEncoderDecoder -> ColumnTypes.Numeric),
    classOf[java.math.BigDecimal] -> (BigDecimalEncoderDecoder -> ColumnTypes.Numeric),

    classOf[java.util.UUID] -> (UUIDEncoderDecoder -> ColumnTypes.UUID),

    classOf[LocalDate] -> ( DateEncoderDecoder -> ColumnTypes.Date ),
    classOf[LocalDateTime] -> (TimestampEncoderDecoder.Instance -> ColumnTypes.Timestamp),
    classOf[DateTime] -> (TimestampWithTimezoneEncoderDecoder -> ColumnTypes.TimestampWithTimezone),
    classOf[ReadableDateTime] -> (TimestampWithTimezoneEncoderDecoder -> ColumnTypes.TimestampWithTimezone),
    classOf[ReadableInstant] -> (DateEncoderDecoder -> ColumnTypes.Date),

    classOf[ReadablePeriod] -> (PostgreSQLIntervalEncoderDecoder -> ColumnTypes.Interval),
    classOf[ReadableDuration] -> (PostgreSQLIntervalEncoderDecoder -> ColumnTypes.Interval),

    classOf[java.util.Date] -> (TimestampWithTimezoneEncoderDecoder -> ColumnTypes.TimestampWithTimezone),
    classOf[java.sql.Date] -> ( DateEncoderDecoder -> ColumnTypes.Date ),
    classOf[java.sql.Time] -> ( SQLTimeEncoder -> ColumnTypes.Time ),
    classOf[java.sql.Timestamp] -> (TimestampWithTimezoneEncoderDecoder -> ColumnTypes.TimestampWithTimezone),
    classOf[java.util.Calendar] -> (TimestampWithTimezoneEncoderDecoder -> ColumnTypes.TimestampWithTimezone),
    classOf[java.util.GregorianCalendar] -> (TimestampWithTimezoneEncoderDecoder -> ColumnTypes.TimestampWithTimezone),
    classOf[Array[Byte]] -> ( ByteArrayEncoderDecoder -> ColumnTypes.ByteA ),
    classOf[ByteBuffer] -> ( ByteArrayEncoderDecoder -> ColumnTypes.ByteA ),
    classOf[ByteBuf] -> ( ByteArrayEncoderDecoder -> ColumnTypes.ByteA )
  )

  private final val classesSequence = (classOf[LocalTime] -> (TimeEncoderDecoder.Instance -> ColumnTypes.Time)) ::
    (classOf[ReadablePartial] -> (TimeEncoderDecoder.Instance -> ColumnTypes.Time)) ::
    classesSequence_

  private final val classes = classesSequence.toMap

  override def encode(value: Any): String = {

    if (value == null) {
      return null
    }

    value match {
      case Some(v) => encode(v)
      case None => null
      case _ => encodeValue(value)
    }

  }

  /**
   * Used to encode a value that is not null and not an Option.
   */
  private def encodeValue(value: Any): String = {

    val encoder = this.classes.get(value.getClass)

    if (encoder.isDefined) {
      encoder.get._1.encode(value)
    } else {

      val view: Option[Traversable[Any]] = value match {
        case i: java.lang.Iterable[_] => Some(i.toIterable)
        case i: Traversable[_] => Some(i)
        case i: Array[_] => Some(i.toIterable)
        case _ => None
      }

      view match {
        case Some(collection) => encodeArray(collection)
        case None => {
          this.classesSequence.find(entry => entry._1.isAssignableFrom(value.getClass)) match {
            case Some(parent) => parent._2._1.encode(value)
            case None => value.toString
          }
        }
      }

    }

  }

  private def encodeArray(collection: Traversable[_]): String = {
    val builder = new StringBuilder()

    builder.append('{')

    val result = collection.map {
      item =>

        if (item == null || item == None) {
          "NULL"
        } else {
          if (this.shouldQuote(item)) {
            "\"" + this.encode(item).replaceAllLiterally("\\", """\\""").replaceAllLiterally("\"", """\"""") + "\""
          } else {
            this.encode(item)
          }
        }

    }.mkString(",")

    builder.append(result)
    builder.append('}')

    builder.toString()
  }

  private def shouldQuote(value: Any): Boolean = {
    value match {
      case n: java.lang.Number => false
      case n: Int => false
      case n: Short => false
      case n: Long => false
      case n: Float => false
      case n: Double => false
      case n: java.lang.Iterable[_] => false
      case n: Traversable[_] => false
      case n: Array[_] => false
      case Some(v) => shouldQuote(v)
      case _ => true
    }
  }

  override def kindOf(value: Any): Int = {
    if ( value == null || value == None ) {
      0
    } else {
      value match {
        case Some(v) => kindOf(v)
        case v : String => ColumnTypes.Untyped
        case _ => {
          this.classes.get(value.getClass) match {
            case Some( entry ) => entry._2
            case None => ColumnTypes.Untyped
          }
        }
      }
    }
  }

}

File: postgresql-async/src/main/java/com/github/mauricio/async/db/postgresql/column/PostgreSQLColumnEncoderRegistry.kt
package com.github.mauricio.async.db.postgresql.column

import com.github.jasync.sql.db.column.BigDecimalEncoderDecoder
import com.github.jasync.sql.db.column.ColumnEncoderRegistry
import com.github.jasync.sql.db.column.DateEncoderDecoder
import com.github.jasync.sql.db.column.DoubleEncoderDecoder
import com.github.jasync.sql.db.column.FloatEncoderDecoder
import com.github.jasync.sql.db.column.InetAddressEncoderDecoder
import com.github.jasync.sql.db.column.IntegerEncoderDecoder
import com.github.jasync.sql.db.column.LongEncoderDecoder
import com.github.jasync.sql.db.column.SQLTimeEncoder
import com.github.jasync.sql.db.column.ShortEncoderDecoder
import com.github.jasync.sql.db.column.StringEncoderDecoder
import com.github.jasync.sql.db.column.TimeEncoderDecoder
import com.github.jasync.sql.db.column.TimestampEncoderDecoder
import com.github.jasync.sql.db.column.TimestampWithTimezoneEncoderDecoder
import com.github.jasync.sql.db.column.UUIDEncoderDecoder
import io.netty.buffer.ByteBuf
import org.joda.time.DateTime
import org.joda.time.LocalDate
import org.joda.time.LocalDateTime
import org.joda.time.LocalTime
import org.joda.time.ReadableDateTime
import org.joda.time.ReadableDuration
import org.joda.time.ReadableInstant
import org.joda.time.ReadablePeriod
import java.math.BigDecimal
import java.nio.ByteBuffer


class PostgreSQLColumnEncoderRegistry : ColumnEncoderRegistry {

  companion object {
    val Instance = PostgreSQLColumnEncoderRegistry()
  }

  private val classesSequence_
    get() = listOf(
        Int.javaClass to (IntegerEncoderDecoder to ColumnTypes.Numeric),
        java.lang.Integer::class.java to (IntegerEncoderDecoder to ColumnTypes.Numeric),

        java.lang.Short::class.java to (ShortEncoderDecoder to ColumnTypes.Numeric),
        Short.javaClass to (ShortEncoderDecoder to ColumnTypes.Numeric),

        Long.javaClass to (LongEncoderDecoder to ColumnTypes.Numeric),
        java.lang.Long::class.java to (LongEncoderDecoder to ColumnTypes.Numeric),

        String.javaClass to (StringEncoderDecoder to ColumnTypes.Varchar),
        java.lang.String::class.java to (StringEncoderDecoder to ColumnTypes.Varchar),

        Float.javaClass to (FloatEncoderDecoder to ColumnTypes.Numeric),
        java.lang.Float::class.java to (FloatEncoderDecoder to ColumnTypes.Numeric),

        Double.javaClass to (DoubleEncoderDecoder to ColumnTypes.Numeric),
        java.lang.Double::class.java to (DoubleEncoderDecoder to ColumnTypes.Numeric),

        BigDecimal::class.java to (BigDecimalEncoderDecoder to ColumnTypes.Numeric),

        java.net.InetAddress::class.java to (InetAddressEncoderDecoder to ColumnTypes.Inet),

        java.util.UUID::class.java to (UUIDEncoderDecoder to ColumnTypes.UUID),

        LocalDate::class.java to (DateEncoderDecoder to ColumnTypes.Date),
        LocalDateTime::class.java to (TimestampEncoderDecoder.Instance to ColumnTypes.Timestamp),
        DateTime::class.java to (TimestampWithTimezoneEncoderDecoder to ColumnTypes.TimestampWithTimezone),
        ReadableDateTime::class.java to (TimestampWithTimezoneEncoderDecoder to ColumnTypes.TimestampWithTimezone),
        ReadableInstant::class.java to (DateEncoderDecoder to ColumnTypes.Date),

        ReadablePeriod::class.java to (PostgreSQLIntervalEncoderDecoder to ColumnTypes.Interval),
        ReadableDuration::class.java to (PostgreSQLIntervalEncoderDecoder to ColumnTypes.Interval),

        java.util.Date::class.java to (TimestampWithTimezoneEncoderDecoder to ColumnTypes.TimestampWithTimezone),
        java.sql.Date::class.java to (DateEncoderDecoder to ColumnTypes.Date),
        java.sql.Time::class.java to (SQLTimeEncoder to ColumnTypes.Time),
        java.sql.Timestamp::class.java to (TimestampWithTimezoneEncoderDecoder to ColumnTypes.TimestampWithTimezone),
        java.util.Calendar::class.java to (TimestampWithTimezoneEncoderDecoder to ColumnTypes.TimestampWithTimezone),
        java.util.GregorianCalendar::class.java to (TimestampWithTimezoneEncoderDecoder to ColumnTypes.TimestampWithTimezone),
        arrayOf<Byte>().javaClass to (ByteArrayEncoderDecoder to ColumnTypes.ByteA),
        ByteBuffer::class.java to (ByteArrayEncoderDecoder to ColumnTypes.ByteA),
        ByteBuf::class.java to (ByteArrayEncoderDecoder to ColumnTypes.ByteA)
    )

  private val classesSequence = LocalTime::class.java to (TimeEncoderDecoder.Instance -> ColumnTypes.Time)
  (ReadablePartial.javaClass -> (TimeEncoderDecoder.Instance -> ColumnTypes.Time))
  classesSequence_

  private val classes = classesSequence.toMap

  override fun encode(value: Any): String {
    if (value == null) {
      return null
    }

     when(value) {
      Some(v) -> encode(v)
      None -> null
      else -> encodeValue(value)
    }

  }

  /**
   * Used to encode a value that is not null and not an Option.
   */
  private fun encodeValue(value: Any): String {

    val encoder = this.classes.get(value.getClass)

    if (encoder.isDefined) {
      encoder.get._1.encode(value)
    } else {
      value when {
        i: java.lang.Iterable<_>
        -> encodeArray(i.toIterable)
        i: Traversable<_>
        -> encodeArray(i)
        i: Array<_>
        -> encodeArray(i.toIterable)
        p: Product
        -> encodeComposite(p)
        else -> {
          this.classesSequence.find(entry -> entry._1.isAssignableFrom(value.getClass)) when {
            Some(parent) -> parent._2._1.encode(value)
            None -> value.toString
          }
        }
      }

    }

  }

  private fun encodeComposite(p: Product): String {
    p.productIterator.map { item ->
      if (item == null || item == None) {
        "NULL"
      } else {
        if (this.shouldQuote(item)) {
          "\"" + this.encode(item).replaceAllLiterally("\\", """\\""").replaceAllLiterally("\"", """\"""") + "\""
        } else {
          this.encode(item)
        }
      }
    }.mkString("(", ",", ")")
  }

  private fun encodeArray(collection: Traversable<_>): String {
    collection.map { item ->
      if (item == null || item == None) {
        "NULL"
      } else {
        if (this.shouldQuote(item)) {
          "\"" + this.encode(item).replaceAllLiterally("\\", """\\""").replaceAllLiterally("\"", """\"""") + "\""
        } else {
          this.encode(item)
        }
      }
    }.mkString("{", ",", "}")
  }

  private fun shouldQuote(value: Any): Boolean {
    value when {
      n: java.lang.Number
      -> false
      n: Int
      -> false
      n: Short
      -> false
      n: Long
      -> false
      n: Float
      -> false
      n: Double
      -> false
      n: java.lang.Iterable<_>
      -> false
      n: Traversable<_>
      -> false
      n: Array<_>
      -> false
      Some(v) -> shouldQuote(v)
      else -> true
    }
  }

  override fun kindOf(value: Any): Int {
    if (value == null || value == None) {
      0
    } else {
      value when {
        Some(v) -> kindOf(v)
        v : String
        -> ColumnTypes.Untyped
        else -> {
          this.classes.get(value.getClass) when {
            Some(entry) -> entry._2
            None -> ColumnTypes.Untyped
          }
        }
      }
    }
  }

}